crates/mirror-mirror-macros/src/derive_reflect/struct_named.rs

```diff
@@ -368,45 +368,83 @@ fn expand_struct(
         }
     };
 
-    let fn_fields = {
-        let crate_name = &attrs.crate_name;
+    let fields_len = fields
+        .iter()
+        .filter(field_attrs.filter_out_skipped_named())
+        .count();
 
+    let fn_fields_len = {
         quote! {
-            fn fields(&self) -> #crate_name::struct_::Iter<'_> {
-                #crate_name::struct_::Iter::new(self)
+            fn fields_len(&self) -> usize {
+                #fields_len
             }
         }
     };
 
-    let fn_fields_mut = {
+    let fn_fields = {
+        let crate_name = &attrs.crate_name;
         let code_for_fields = fields
             .iter()
             .filter(field_attrs.filter_out_skipped_named())
             .map(|field| {
                 let ident = &field.ident;
                 let field = stringify(ident);
                 quote! {
-                    (#field, self.#ident.as_reflect_mut()),
+                    (#field, self.#ident.as_reflect()),
                 }
             });
 
-        quote! {
-            fn fields_mut(&mut self) -> PairIterMut<'_> {
-                let iter = [#(#code_for_fields)*];
-                Box::new(iter.into_iter())
+        if fields_len <= 32 {
+            quote! {
+                fn fields(&self) -> #crate_name::struct_::FieldsIter<'_> {
+                    let fields = [
+                        #(#code_for_fields)*
+                    ];
+                    Box::new(fields.into_iter())
+                }
+            }
+        } else {
+            quote! {
+                fn fields(&self) -> #crate_name::struct_::FieldsIter<'_> {
+                    let fields = vec![
+                        #(#code_for_fields)*
+                    ];
+                    Box::new(fields.into_iter())
```

Comment:
A little unfortunate for the non-mutating case to require a box. Why is that?

Reply:
I believe this will actually be faster than the previous version. Box is a little unfortunate indeed but the reason is I wanted to be able to specialize the iterator type per implementation. Previously it didn't allocate but on each `next()` call there was quite a lot to do. Now we box, but each call to `next()` should be trivial. Which I think will actually win out.